zerver/lib/api.py

```diff
@@ -0,0 +1,334 @@
+    ParamSpec,
+    get_args,
+    get_origin,
+    get_type_hints,
+)
+
+from zerver.lib.converters import format_exception, json_converter, zulip_converter
+from zerver.lib.exceptions import InvalidJSONError, JsonableError
+from zerver.lib.request import RequestNotes, RequestVariableMissingError, arguments_map
+from zerver.lib.response import MutableJsonResponse
+
+T = TypeVar("T")
+ParamT = ParamSpec("ParamT")
+ReturnT = TypeVar("ReturnT")
+
+
+class DocumentationStatus(Enum):
+    DOCUMENTED = auto()
+    INTENTIONALLY_UNDOCUMENTED = auto()
+    DOCUMENTATION_PENDING = auto()
+
+
+@frozen
+class Metadata:
+    whence: Optional[str] = None
+    path_only: bool = False
+    convert_from_json: bool = False
+    argument_type_is_body: bool = field(default=False)
+    documentation_status: DocumentationStatus = DocumentationStatus.DOCUMENTED
+
+    @argument_type_is_body.validator
+    def _check(self, attribute: object, argument_type_is_body: bool) -> None:
+        if argument_type_is_body:
+            assert (
+                self.whence is None
+            ), "argument_type_is_body is incompatible with a non-default whence"
+
+
+# TypeAliases for common Annotated types
+
+JSON = Annotated[T, Metadata(convert_from_json=True)]
+WebhookPayload = Annotated[T, Metadata(convert_from_json=True, argument_type_is_body=True)]
+PathOnly = Annotated[T, Metadata(path_only=True)]
+
+# Implementation
+
+
+class _NotSpecified:
+    pass
+
+
+NotSpecified = _NotSpecified()
+
+
+@frozen
+class FuncParam(Generic[T]):
+    # Metadata retrieved from the Annotated type.
+    metadata: Metadata
+    # Name of the function parameter as defined in the original function.
+    param_name: str
+    # Inspected underlying type of the parameter unwarping the Annotated type if
+    # there is one.
+    param_type: Type[T]
+    # Name of the corresponding variable in the request data to look for. When
+    # argument_type_is_body is True, this is set to "request".
+    request_var_name: str
+    # Default value of the parameter.
+    default: Union[T, _NotSpecified] = NotSpecified
+
+
+@frozen
+class ViewFuncInfo:
+    view_func_full_name: str
+    parameters: Sequence[FuncParam[object]]
+    # Whether a request is expected to be processed by another function
+    # decorated as an endpoint or not.
+    no_nesting: bool
+
+
+@frozen
+class CallInfo:
+    kwargs_to_update: Mapping[str, object]
+    processed_request_var_names: Set[str]
+    errors: Sequence[str]
+
+
+IGNORED_PARAMS = frozenset(["request", "user_profile", "args", "kwargs"])
+
+
+def parse_view_func_signature(
+    view_func: Callable[Concatenate[HttpRequest, ParamT], object]
+) -> ViewFuncInfo:
+    """This is responsible for inspecting the function signature and getting the
+    metadata from the parameters. We want to keep this function as pure as
+    possible not leaking side effects to the global state. Side effects should
+    be emitted separately after the ViewFuncInfo is returned.
```

Comment:
emitted -> executed? (The word "emit" feels a little funny here, but I could be wrong. Feel free to disregard.)